### PR TITLE
Move to `criterion` benchmarks

### PR DESCRIPTION
**Pull request by Pratyush**
## Description

Moves from the old `bencher` based infrastructure to a more accurate `criterion` based infrastructure. This allows us to minimize the time spent in generating material for initializing field and group elements, and also enables us to add better profilers in the future.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the GitHub PR explorer

### CODE REVIEW

File: bench-templates/src/macros/field.rs [hunk 1]
```diff
@@ -1,542 +1,468 @@
...
+                    let mut i = 0;
+                    b.iter(|| {
+                        i = (i + 1) % SAMPLES;
+                        let j = (i + 1) % SAMPLES;
+                        field_elements_left[i] * field_elements_left[j]
+                            + field_elements_right[j] * field_elements_right[i]
+                    })
+                },
+            );
+            // arithmetic.bench_function(
```
Comment (weikengchen) [inline bench-templates/src/macros/field.rs:146]: What is the plan?

Comment (Pratyush) [reply bench-templates/src/macros/field.rs:146]: Oh oops, good catch. Adding these back now.

File: bench-templates/src/macros/field.rs [hunk 2]
```diff
@@ -1,542 +1,468 @@
...
-            let v: Vec<$f_type> = (0..SAMPLES)
-                .map(|_| {
-                    let mut tmp = $f::rand(&mut rng);
-                    tmp.square_in_place();
-                    tmp
+            let f = (0..SAMPLES)
+                .map(|_| <$F>::rand(&mut rng))
+                .collect::<Vec<_>>();
+            let qrs = f.iter().map(|s| s.square()).collect::<Vec<_>>();
+            let qnrs = f.iter().map(|s| s.square()).collect::<Vec<_>>();
```
Comment (weikengchen) [inline bench-templates/src/macros/field.rs:274]: Would `qnrs` be NRS?

File: bench-templates/src/macros/field.rs [hunk 3]
```diff
@@ -1,542 +1,468 @@
...
+                    v1[i] == v2[i]
+                })
             });
         }
 
-        fn into_repr(b: &mut $crate::bencher::Bencher) {
-            const SAMPLES: usize = 1000;
-
-            let mut rng = ark_std::test_rng();
+        // fn into_repr(b: &mut $crate::bencher::Bencher) {
```
Comment (weikengchen) [inline bench-templates/src/macros/field.rs:438]: Plan?

**Review by weikengchen — APPROVED** [review]

File: bench-templates/src/macros/field.rs (hunk 2 above)
Comment (Pratyush) [reply bench-templates/src/macros/field.rs:274]: `qrs` = quadratic residues, and `qnrs` = quadratic non-residues. I think this is in line with our terminology elsewhere.

Comment (weikengchen) [reply bench-templates/src/macros/field.rs:274]: The problem is that
```
 let qnrs = f.iter().map(|s| s.square()).collect::<Vec<_>>();
```
here qnrs are also qrs?

Comment (Pratyush) [reply bench-templates/src/macros/field.rs:274]: I see. I was planning originally on making it `generator * s.square()`, but the generator doesn't yet exist for all fields. So for the time being I will remove it, as I don't know of another way to easily generate a non-square.

Comment (weikengchen) [reply bench-templates/src/macros/field.rs]: Is it true that for a lot of the field we have access to a QNRS?

Comment (Pratyush) [reply bench-templates/src/macros/field.rs]: Yeah, for Fp the generator is always a QNR, but for extension fields we don't necessarily have QNRs

File: bench-templates/src/macros/field.rs (hunk 3 above)
Comment (Pratyush) [reply bench-templates/src/macros/field.rs]: Fixed in [`b0e862b` (#472)](https://github.com/arkworks-rs/algebra/pull/472/commits/b0e862b405e3114ea353432a4c273defbb7e7c88)

File: bench-templates/src/macros/field.rs (hunk 1 above)
Comment (Pratyush) [reply bench-templates/src/macros/field.rs]: Fixed in [`b0e862b` (#472)](https://github.com/arkworks-rs/algebra/pull/472/commits/b0e862b405e3114ea353432a4c273defbb7e7c88)

File: bench-templates/src/macros/field.rs (hunk 2 above)
Comment (weikengchen) [reply bench-templates/src/macros/field.rs]: I feel maybe one day we will add a mandatory QNR field for extension fields. Thoughts?

Comment (Pratyush) [reply bench-templates/src/macros/field.rs]: This is related to #276

Comment (mmagician) [reply bench-templates/src/macros/field.rs]: Some of the extension fields used to have QNRs which were not used anywhere, see https://github.com/arkworks-rs/algebra/pull/421/.

Comment (Pratyush) [reply bench-templates/src/macros/field.rs]: Hm good catch, still not sure how we could use this generically.

**Pull request merged**